crypto/kdf/sshkdf.c

```diff
@@ -0,0 +1,285 @@
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <openssl/evp_kdf.h>
+#include <openssl/evp.h>
+#include "internal/cryptlib.h"
+#include "internal/evp_int.h"
+
+/* See RFC 4253, Section 7.2 */
+
+static int SSHKDF(const EVP_MD *evp_md,
+                  const unsigned char *key, size_t key_len,
+                  const unsigned char *xcghash, size_t xcghash_len,
+                  const unsigned char *session_id, size_t session_id_len,
+                  char type, unsigned char *okey, size_t okey_len);
+
+typedef struct {
+    const EVP_MD *md;
+    unsigned char *key; /* K */
+    size_t key_len;
+    unsigned char *xcghash; /* H */
+    size_t xcghash_len;
+    char type; /* X */
+    unsigned char *session_id;
+    size_t session_id_len;
+} SSHKDF_KDF_CTX;
+
+static int kdf_sshkdf_init(EVP_KDF_CTX *ctx)
+{
+    SSHKDF_KDF_CTX *kctx;
+
+    if ((kctx = OPENSSL_zalloc(sizeof(*kctx))) == NULL) {
+        KDFerr(KDF_F_KDF_SSHKDF_INIT, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    ctx->data = kctx;
+
+    return 1;
+}
+
+static void kdf_sshkdf_cleanup(EVP_KDF_CTX *ctx)
+{
+    SSHKDF_KDF_CTX *kctx = ctx->data;
+    OPENSSL_clear_free(kctx->key, kctx->key_len);
+    OPENSSL_clear_free(kctx->xcghash, kctx->xcghash_len);
+    OPENSSL_clear_free(kctx->session_id, kctx->session_id_len);
+    OPENSSL_free(kctx);
+}
+
+static int kdf_sshkdf_ctrl(EVP_KDF_CTX *ctx, int type, int p1, void *p2)
+{
+    SSHKDF_KDF_CTX *kctx = ctx->data;
+
+    switch (type) {
+    case EVP_KDF_CTRL_MD:
+        if (p2 == NULL)
+            return 0;
+
+        kctx->md = p2;
+        return 1;
+
+    case EVP_KDF_CTRL_KEY:
+        if (p1 < 0)
+            return 0;
+
+        if (kctx->key != NULL)
+            OPENSSL_clear_free(kctx->key, kctx->key_len);
+
+        kctx->key = OPENSSL_memdup(p2, p1);
+        if (kctx->key == NULL)
+            return 0;
+
+        kctx->key_len  = p1;
+        return 1;
+
+    case EVP_KDF_CTRL_SSHKDF_XCGHASH:
+        if (p1 < 0)
+            return 0;
+
+        if (kctx->xcghash != NULL)
+            OPENSSL_clear_free(kctx->xcghash, kctx->xcghash_len);
+
+        kctx->xcghash = OPENSSL_memdup(p2, p1);
+        if (kctx->xcghash == NULL)
+            return 0;
+
+        kctx->xcghash_len  = p1;
+        return 1;
+
+    case EVP_KDF_CTRL_SSHKDF_SESSION_ID:
+        if (p1 < 0)
+            return 0;
+
+        if (kctx->session_id != NULL)
+            OPENSSL_clear_free(kctx->session_id, kctx->session_id_len);
+
+        kctx->session_id = OPENSSL_memdup(p2, p1);
+        if (kctx->session_id == NULL)
+            return 0;
+
+        kctx->session_id_len  = p1;
+        return 1;
+
+    case EVP_KDF_CTRL_SSHKDF_TYPE:
+        if (p1 < (int)'A' || p1 > (int)'F') {
+            KDFerr(KDF_F_KDF_SSHKDF_CTRL_STR, KDF_R_VALUE_ERROR);
```

Comment:
`KDF_F_KDF_SSHKDF_CTRL_STR` should actually be `KDF_F_KDF_SSHKDF_CTRL`

This is the cause of current Travis failure.

Reply:
Ha, odd, all tests passed here, travis was still running last night, I'll fix

Reply:
Are the A and F char constants right?  Should it be using a ctype macro?  Does it work on ebcdic, etc., codesets?

Reply:
A to F are constants, yes, the RFC does not defined them very clearly but mentions `"A" means the single character A, ASCII 65`, it is a single byte in the KDF.
What would a "ctype macro" lok like ?

Reply:
@richsalz  changed the check to use integers so it won't depend on charset
I also added 6 macros with more useful names (from the RFC), the macros are not necessary, so you tell me if we want to keep them or not.